### PR TITLE
BTreeMap binary serialization

### PR DESCRIPTION
**Pull request by NikVolf**

### CODE REVIEW

File: ipc/rpc/src/binary.rs [hunk 1]
```diff
@@ -139,6 +139,96 @@ impl<R: BinaryConvertable, E: BinaryConvertable> BinaryConvertable for Result<R,
...
+			if val_size > 0 {
+				let item_end = offset + key_size;
+				try!(val.to_bytes(&mut buffer[offset..item_end], length_stack));
+				offset = item_end;
+			}
+		}
+		Ok(())
+	}
+
+	fn from_bytes(buffer: &[u8], length_stack: &mut VecDeque<usize> ) -> Result<Self, BinaryConvertError> {
```
Comment (gavofyork) [inline ipc/rpc/src/binary.rs]: superfluous space.

File: ipc/rpc/src/binary.rs [hunk 2]
```diff
@@ -139,6 +139,96 @@ impl<R: BinaryConvertable, E: BinaryConvertable> BinaryConvertable for Result<R,
...
+		if buffer.len() == 0 { return Ok(result); }
+
+		loop {
+			let key_size = match K::len_params() {
+				0 => mem::size_of::<K>(),
+				_ => try!(length_stack.pop_front().ok_or(BinaryConvertError)),
+			};
+			let key = if key_size == 0 {
+				try!(K::from_empty_bytes())
+			}
```
Comment (gavofyork) [inline ipc/rpc/src/binary.rs]: convention is `} else {` on same line

File: ipc/rpc/src/binary.rs [hunk 3]
```diff
@@ -139,6 +139,96 @@ impl<R: BinaryConvertable, E: BinaryConvertable> BinaryConvertable for Result<R,
...
+			index = index + key_size;
+
+			let val_size = match V::len_params() {
+				0 => mem::size_of::<V>(),
+				_ => try!(length_stack.pop_front().ok_or(BinaryConvertError)),
+			};
+			let val = if val_size == 0 {
+				try!(V::from_empty_bytes())
+			}
+			else {
```
Comment (gavofyork) [inline ipc/rpc/src/binary.rs]: same as above

File: ipc/rpc/src/binary.rs [hunk 4]
```diff
@@ -139,6 +139,96 @@ impl<R: BinaryConvertable, E: BinaryConvertable> BinaryConvertable for Result<R,
 	}
 }
 
+impl<K, V> BinaryConvertable for BTreeMap<K, V> where K : BinaryConvertable + Ord, V: BinaryConvertable {
+	fn size(&self) -> usize {
+		0usize + match K::len_params() {
```
Comment (gavofyork) [inline ipc/rpc/src/binary.rs]: indentation less than great. would suggest:

```
0usize + match K::len_params() {
		0 => mem::size_of::<K>() * self.len(),
		_ => self.iter().fold(0usize, |acc, (k, _)| acc + k.size())
} + match V::len_params() {
		0 => mem::size_of::<V>() * self.len(),
		_ => self.iter().fold(0usize, |acc, (_, v)| acc + v.size())
}
```

**Comment by gavofyork** [comment]
minor stylist grumbles. good to merge otherwise.

**Comment by NikVolf** [comment]
@gavofyork fixed

**Pull request merged**